src/calculator/index.ts

```diff
@@ -321,24 +323,27 @@ export default class Calculator {
 
         /*TODO: take into account round matching token decimals instead of default 18 */
         const amountRoundToken =
-          (Number(amountUSD) / conversionRateRoundToken.price) *
-          Math.pow(10, 18);
+          (Number(amountUSD) / conversionRateRoundToken.price) * 10e18;
 
         return {
           ...vote,
           amount: vote.amount.toString(),
-          amountRoundToken: Math.floor(amountRoundToken).toString(),
+          amountRoundToken: BigInt(Math.floor(amountRoundToken)).toString(),
```

Comment:
you can use .toLocaleString("fullwide", { useGrouping: false }) for that.
otherwise we should convert the numbers before the multiplication that make them big in case they become too big for numbers

Reply:
Relying on BigInt's toString feels better than relying on a toLocaleString, but I'll change it. Let's just get this merged asap, this is a blocker for testing and shipping matching estimates.

Reply:
I'm gonna make the call to rely on the standard BigInt toString instead of the toLocaleString, because it's clearer what the code is doing. requesting approval @boudra @gravityblast @bard @bhargavaparoksham

Reply:
+1 from me.

Reply:
@vacekj looking at this code closer, I think there are a couple of places where we could lose precision:

1. `formatUnits(vote.amount, 18)` converting the bigint to a number string here and parsing it again as a number is unnecessarily risky
2. `const amountRoundToken = (Number(amountUSD) / conversionRateRoundToken.price) * 10e18;` this multiplication will easily result a higher number than `Number.MAX_SAFE_INTEGER` (9.07*10e15), leading to possible loss of precision
3. `Math.floor(amountRoundToken)` is also a potential loss of precision
4. We're not taking into account the token decimals, this would be fine if we didn't have support USDC which has different decimals than the rest

not sure if you saw how we're doing the conversion here: https://github.com/gitcoinco/allo-indexer/blob/main/src/prices/provider.ts#L104

in that calculation, we truncate the USD price to 8 places and then do all the math in big integers, it also takes into account the token decimals

this could all still be working fine, but if we already have a tested calculation in place, I would try to use that instead of coming up with new less precise calculation, that calculation can be split up into it's own function if needed

Reply:
@vacekj @boudra would the above be reason enough to have a go-to `token-math.ts` module (or something like that) where this sort of calculations is centralized and, more importantly, tested against edge cases?

Reply:
@bard This type of math can get complex easily so I think it's the way to go. I would already do it in this PR, off the top of my head I think all we would have to do is split up the logic that already exists into a pure functions in `tokenMath.ts` and implement some tests for it, I'd be more than happy to help here @vacekj since I implemented those first calculations.

Reply:
I also noticed that those calculations have a lot of math in one line expressions so they might be hard to read, again happy to help here!